lightning/src/util/sweep.rs

```diff
@@ -0,0 +1,675 @@
+			entropy_source,
+			kv_store,
+			best_block,
+			chain_data_source,
+			logger,
+			spend_outputs_callback,
+			regenerate_spend_threshold,
+		}
+	}
+
+	/// Tells the sweeper to track the given outputs descriptors.
+	///
+	/// Usually, this should be called based on the values emitted by the
+	/// [`Event::SpendableOutputs`] event.
+	///
+	/// The given `exclude_static_ouputs` flag controls whether the sweeper will filter out
+	/// [`SpendableOutputDescriptor::StaticOutput`]s that might be handled directly by the on-chain
+	/// wallet implementation.
+	///
+	/// [`Event::SpendableOutputs`]: crate::events::Event::SpendableOutputs
+	pub fn track_spendable_outputs(
+		&self, mut output_descriptors: Vec<SpendableOutputDescriptor>,
+		channel_id: Option<ChannelId>, exclude_static_ouputs: bool,
+	) {
+		let relevant_descriptors = if exclude_static_ouputs {
+			output_descriptors
+				.drain(..)
+				.filter(|desc| !matches!(desc, SpendableOutputDescriptor::StaticOutput { .. }))
+				.collect::<Vec<_>>()
+		} else {
+			output_descriptors
+		};
+
+		if relevant_descriptors.is_empty() {
+			return;
+		}
+
+		{
+			let mut locked_outputs = self.outputs.lock().unwrap();
+			for descriptor in relevant_descriptors {
+				let id = self.entropy_source.get_secure_random_bytes();
+				let output_info = TrackedSpendableOutput {
+					id,
+					descriptor,
+					channel_id,
+					first_broadcast_hash: None,
+					latest_broadcast_height: None,
+					latest_spending_tx: None,
+					confirmation_height: None,
+					confirmation_hash: None,
+				};
+
+				locked_outputs.push(output_info.clone());
+				self.persist_info(&output_info).unwrap_or_else(|e| {
+					log_error!(self.logger, "Error persisting TrackedSpendableOutput: {:?}", e);
+				});
+			}
+		}
+
+		self.rebroadcast_if_necessary();
+	}
+
+	/// Returns a list of the currently tracked spendable outputs.
+	pub fn tracked_spendable_outputs(&self) -> Vec<TrackedSpendableOutput> {
+		self.outputs.lock().unwrap().clone()
+	}
+
+	fn rebroadcast_if_necessary(&self) {
+		let (cur_height, cur_hash) = {
+			let best_block = self.best_block.lock().unwrap();
+			(best_block.height(), best_block.block_hash())
+		};
+
+		let mut respend_descriptors = Vec::new();
+		let mut respend_ids = Vec::new();
+
+		{
+			let mut locked_outputs = self.outputs.lock().unwrap();
+			for output_info in locked_outputs.iter_mut() {
+				if output_info.confirmation_height.is_some() {
+					// Don't rebroadcast confirmed txs
+					debug_assert!(output_info.confirmation_hash.is_some());
+					continue;
+				}
+
+				if let Some(latest_broadcast_height) = output_info.latest_broadcast_height {
+					// Re-generate spending tx after regenerate_spend_threshold, rebroadcast
+					// after every block
+					if latest_broadcast_height + self.regenerate_spend_threshold >= cur_height {
+						respend_descriptors.push(output_info.descriptor.clone());
+						respend_ids.push(output_info.id);
+					} else if latest_broadcast_height < cur_height {
```

Comment:
not entirly sure the `if` here is needed, as if the previous `if` is `false` then `cur_height > latest_broadcast_height + self.regenerate_spend_threshold` and `self.regenerate_Spend_threshold` is `> 0`